src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs

```diff
@@ -104,42 +152,77 @@ public CloudEvent DecodeJObject(JObject jObject, IEnumerable<CloudEventAttribute
+            {
+                throw new ArgumentException($"Structured mode content cannot contain both '{Data}' and '{DataBase64}' properties.");
+            }
+            if (dataBase64Token is object)
+            {
+                if (dataBase64Token.Type != JTokenType.String)
+                {
+                    throw new ArgumentException($"Structured mode property '{DataBase64}' must be a string, when present.");
+                }
+                cloudEvent.Data = Convert.FromBase64String((string) dataBase64Token);
+            }
+            else
+            {
+                if (dataToken.Type == JTokenType.Null)
+                {
+                    return;
+                }
+                // Convert JSON string tokens to string values when the content type suggests that's appropriate,
+                // otherwise leave the token as it is.
+                cloudEvent.Data = dataToken.Type == JTokenType.String && cloudEvent.DataContentType?.StartsWith("text/") == true
+                    ? (string) dataToken
+                    : (object) dataToken; // Deliberately cast to object to avoid any implicit conversions
+            }
         }
 
-        public byte[] EncodeStructuredEvent(CloudEvent cloudEvent, out ContentType contentType)
+        public override byte[] EncodeStructuredModeMessage(CloudEvent cloudEvent, out ContentType contentType)
         {
             contentType = new ContentType("application/cloudevents+json")
             {
                 CharSet = Encoding.UTF8.WebName
             };
 
             JObject jObject = new JObject();
+            jObject[CloudEventsSpecVersion.SpecVersionAttribute.Name] = cloudEvent.SpecVersion.VersionId;
             var attributes = cloudEvent.GetPopulatedAttributes();
             foreach (var keyValuePair in attributes)
             {
                 jObject[keyValuePair.Key.Name] = JToken.FromObject(keyValuePair.Value);
             }
 
-            // FIXME: This is all a bit arbitrary.
             if (cloudEvent.Data is object)
             {
-                // FIXME: This assumes there's nothing beyond the media type...
-                if (cloudEvent.DataContentType == "application/json")
+                ContentType dataContentType = new ContentType(cloudEvent.DataContentType ?? JsonMediaType);
+                if (dataContentType.MediaType == JsonMediaType)
                 {
-                    jObject[Data] = JToken.FromObject(cloudEvent.Data);
+                    jObject[Data] = cloudEvent.Data is JToken token
+                        ? token
+                        : JToken.FromObject(cloudEvent.Data);
                 }
-                else if (cloudEvent.Data is string text && cloudEvent.DataContentType?.StartsWith("text/") == true)
+                else if (cloudEvent.Data is string text && dataContentType.MediaType.StartsWith("text/"))
                 {
                     jObject[Data] = text;
                 }
```

Comment:
Should we be checking for `IEnumerable<ReadOnlyMemory<byte>>`? I think this would also catch byte array.

Reply:
I haven't added a System.Memory dependency yet - that should be the next topic of discussion for Friday. I wanted to try to get the overall responsibility shape sorted first, then we can work out whether every `byte[]` occurrence should actually be `ReadOnlyMemory<byte>`. (Not sure I follow about `IEnumerable<ReadOnlyMemory<byte>>` but we can talk about that.)